### PR TITLE
Mark RNGCryptoServiceProvider as Obsolete

### LINKED ISSUE #40169: API proposal: Obsolete RNGCryptoServiceProvider
## Background and Motivation

The `RNGCryptoServiceProvider` class is a relic from the old Windows CAPI days of yore. The original Win32 API is [no longer recommended](https://docs.microsoft.com/en-us/windows/win32/api/wincrypt/nf-wincrypt-cryptgenrandom), preferring [CNG](https://docs.microsoft.com/en-us/windows/win32/api/bcrypt/nf-bcrypt-bcryptgenrandom) for all new work.

In .NET Core, the `RNGCryptoServiceProvider` type is marked `[EditorBrowsable(Never)]`, and the implementation ignores all provided constructor parameters and delegates to the underlying preferred OS implementation anyway.

There's no reason for an application targeting .NET 6.0+ to use this API. Apps should instead use `RandomNumberGenerator.Create()`. For AOT and linker trimming scenarios, this could also help eliminate the app's dependency on the package which contains the `RNGCryptoServiceProvider` type, reducing overall memory usage and disk footprint.

## Proposed API

```cs
namespace System.Security.Cryptography
{
    [EditorBrowsable(EditorBrowsableState.Never)] // existing attribute
    [Obsolete("This type is obsolete. Use RandomNumberGenerator.Create() instead.")] // new attribute
    public sealed class RNGCryptoServiceProvider : RandomNumberGenerator
    { /* ... */ }
}
```

This could be accompanied by a fixer with two behaviors:

- All calls to `RNGCryptoServiceProvider` ctors become calls to the parameterless overload `RandomNumberGenerator.Create()`.
- All fields / locals / parameters of type `RNGCryptoServiceProvider` instead become type `RandomNumberGenerator`.

The obsoletion __would not__ affect apps targeting netstandard or .NET versions prior to 6.0, as the reference assemblies would not contain these annotations. However, the fixer could apply to all target frameworks.

### PR DESCRIPTION
**Pull request by jeffhandley**
Fixes #40169, marking `RNGCryptoServiceProvider` as obsolete, guiding usage toward `RandomNumberGenerator.Create()`.

### CODE REVIEW

**Comment by jeffhandley** [comment]
The breaking change documentation will be covered by https://github.com/dotnet/docs/issues/23856.

/cc @gewarren

File: docs/project/list-of-diagnostics.md [hunk 1]
```diff
@@ -70,6 +70,7 @@ The PR that reveals the implementation of the `<IncludeInternalObsoleteAttribute
 |  __`SYSLIB0018`__ | ReflectionOnly loading is not supported and throws PlatformNotSupportedException. |
 |  __`SYSLIB0019`__ | RuntimeEnvironment members SystemConfigurationFile, GetRuntimeInterfaceAsIntPtr, and GetRuntimeInterfaceAsObject are no longer supported and throw PlatformNotSupportedException. |
 |  __`SYSLIB0020`__ | JsonSerializerOptions.IgnoreNullValues is obsolete. To ignore null values when serializing, set DefaultIgnoreCondition to JsonIgnoreCondition.WhenWritingNull. |
+|  __`SYSLIB0023`__ | RNGCryptoServiceProvider is obsolete. Use RandomNumberGenerator.Create() instead. |
```
Comment (bartonjs) [inline docs/project/list-of-diagnostics.md:73]: I think the actual guidance we want is to completely abandon the instances and just use the static RandomNumberGenerator.Fill method.

Would you like to assent or rebut, @GrabYourPitchforks?

**Comment by jeffhandley** [comment]
The failing check is #51584.

File: docs/project/list-of-diagnostics.md (hunk 1 above)
Comment (GrabYourPitchforks) [reply docs/project/list-of-diagnostics.md:79]: I agree with Jeremy. The only people who see this message are devs targeting .NET 6, and we should direct all such users to the static methods on `RandomNumberGenerator` instead.

There is one exception: if you're using DI to inject an instance of `RandomNumberGenerator` into your testable classes, you'll want to use `RandomNumberGenerator.Create()` to create an instance and register it as your singleton.

File: src/libraries/System.IO.Pipes/tests/NamedPipeTests/NamedPipeTest.CurrentUserOnly.Windows.cs [hunk 2]
```diff
@@ -22,7 +22,7 @@ public class TestAccountImpersonator : IDisposable
         public TestAccountImpersonator()
         {
             string testAccountPassword;
-            using (RandomNumberGenerator rng = new RNGCryptoServiceProvider())
+            using (RandomNumberGenerator rng = RandomNumberGenerator.Create())
             {
                 var randomBytes = new byte[33];
```
Comment (GrabYourPitchforks) [inline src/libraries/System.IO.Pipes/tests/NamedPipeTests/NamedPipeTest.CurrentUserOnly.Windows.cs:27]: Prefer `var randomBytes = RandomNumberGenerator.GetBytes(33);`, then you can drop lines 25 and 28.

File: docs/project/list-of-diagnostics.md (hunk 1 above)
Comment (jeffhandley) [reply docs/project/list-of-diagnostics.md]: Thanks. Updated; let me know if you have better wording though.

File: src/libraries/System.Security.Principal.Windows/tests/WindowsIdentityImpersonatedTests.netcoreapp.cs [hunk 3]
```diff
@@ -128,7 +128,7 @@ public WindowsTestAccount(string userName)
     private void CreateUser()
     {
         string testAccountPassword;
-        using (RandomNumberGenerator rng = new RNGCryptoServiceProvider())
+        using (RandomNumberGenerator rng = RandomNumberGenerator.Create())
```
Comment (jeffhandley) [inline src/libraries/System.Security.Principal.Windows/tests/WindowsIdentityImpersonatedTests.netcoreapp.cs:131]: I'll update this too

**Review by GrabYourPitchforks — APPROVED** [review]

File: src/libraries/System.IO.Pipes/tests/NamedPipeTests/NamedPipeTest.CurrentUserOnly.Windows.cs [hunk 4]
```diff
@@ -22,14 +22,10 @@ public class TestAccountImpersonator : IDisposable
         public TestAccountImpersonator()
         {
             string testAccountPassword;
-            using (RandomNumberGenerator rng = new RNGCryptoServiceProvider())
-            {
-                var randomBytes = new byte[33];
-                rng.GetBytes(randomBytes);
+            var randomBytes = RandomNumberGenerator.GetBytes(33);
```
Comment (bartonjs) [inline src/libraries/System.IO.Pipes/tests/NamedPipeTests/NamedPipeTest.CurrentUserOnly.Windows.cs:25]: Improper use of `var`:

```suggestion
            byte[] randomBytes = RandomNumberGenerator.GetBytes(33);
```

Comment (jeffhandley) [reply src/libraries/System.IO.Pipes/tests/NamedPipeTests/NamedPipeTest.CurrentUserOnly.Windows.cs]: Thanks for fixing that, @bartonjs.

**Pull request merged**